barretenberg/cpp/src/barretenberg/vm/avm/trace/public_inputs.hpp

```diff
@@ -101,6 +102,32 @@ struct TreeSnapshots {
     AppendOnlyTreeSnapshot note_hash_tree;
     AppendOnlyTreeSnapshot nullifier_tree;
     AppendOnlyTreeSnapshot public_data_tree;
+    inline bool operator==(const TreeSnapshots& rhs) const
+    {
+        return l1_to_l2_message_tree == rhs.l1_to_l2_message_tree && note_hash_tree == rhs.note_hash_tree &&
+               nullifier_tree == rhs.nullifier_tree && public_data_tree == rhs.public_data_tree;
+    }
+    inline TreeSnapshots copy()
+    {
+        return {
+        .l1_to_l2_message_tree = {
+          .root = l1_to_l2_message_tree.root,
+          .size = l1_to_l2_message_tree.size,
+        },
+        .note_hash_tree = {
+          .root = note_hash_tree.root,
+          .size = note_hash_tree.size,
+        },
+        .nullifier_tree = {
+          .root = nullifier_tree.root,
+          .size = nullifier_tree.size,
+        },
+        .public_data_tree = {
+          .root = public_data_tree.root,
+          .size = public_data_tree.size,
+        },
+      };
+    }
```

Comment:
I don't know C++.... Did I need to do this to get a deep copy of a tree snapshot?